core/src/main/java/org/apache/calcite/sql2rel/RelStructuredTypeFlattener.java

```diff
@@ -587,86 +561,103 @@ private void flattenProjections(RewriteRexShuttle shuttle,
-        if (oldOperands.get(0) instanceof RexInputRef) {
-          final RexInputRef inputRef = (RexInputRef) oldOperands.get(0);
-          final Ord<RelDataType> newField =
-              getNewFieldForOldInput(inputRef.getIndex());
-          newExp = rexBuilder.makeCall(exp.getType(),
-              ((RexCall) exp).getOperator(),
-              ImmutableList.of(rexBuilder.makeInputRef(newField.e, newField.i),
-                  oldOperands.get(1)));
-        }
-        for (RelDataTypeField field : newExp.getType().getFieldList()) {
-          flattenedExps.add(
-              Pair.of(rexBuilder.makeFieldAccess(newExp, field.getIndex()),
-                  fieldName + "$" + (j++)));
+        List<RexNode> operands = ((RexCall) exp).getOperands();
+        SqlOperator operator = ((RexCall) exp).getOperator();
+
+        if (operator == SqlStdOperatorTable.ITEM
+            && operands.get(0).getType().isStruct()
+            && operands.get(1).isA(SqlKind.LITERAL)
+            && SqlTypeUtil.inCharFamily(operands.get(1).getType())) {
+          String literalString = ((RexLiteral) operands.get(1)).getValueAs(String.class);
+          RexNode firstOp = operands.get(0);
+
+          if (firstOp instanceof RexInputRef) {
+            // when performed getting field from struct exp by field name
+            // and new input is flattened it's enough to refer target field by index.
+            // But it's possible that requested field is also of type struct, that's
+            // why we're trying to get range from to. For primitive just one field will be in range.
+            int from = 0;
+            for (RelDataTypeField field : firstOp.getType().getFieldList()) {
+              if (literalString.equalsIgnoreCase(field.getName())) {
+                int oldOrdinal = ((RexInputRef) firstOp).getIndex();
+                int to = from + postFlattenSize(field.getType());
+                for (int newInnerOrdinal = from; newInnerOrdinal < to; newInnerOrdinal++) {
+                  Ord<RelDataType> newField = getNewFieldForOldInput(oldOrdinal, newInnerOrdinal);
+                  RexInputRef newRef = rexBuilder.makeInputRef(newField.e, newField.i);
+                  flattenedExps.add(Pair.of(newRef, fieldName));
+                }
+                break;
+              } else {
+                from += postFlattenSize(field.getType());
+              }
+            }
+          } else if (firstOp instanceof RexCall) {
+            // to get nested struct from return type of firstOp rex call,
+            // we need to flatten firstOp and get range of expressions which
+            // corresponding to desirable nested struct flattened fields
+            List<Pair<RexNode, String>> firstOpFlattenedExps = new ArrayList<>();
+            flattenProjection(shuttle, firstOp, fieldName, firstOpFlattenedExps);
+            int newInnerOrdinal = getNewInnerOrdinal(firstOp, literalString);
```

Comment:
It does not seem right we still passed in the `fieldName ` which is the name of the `exp` but not the `firstOp` which does not have a explicit name.

Reply:
I think passing old name here is better for preserving some history, although if you have reasonable alternative, I won't mind changing the code.

Reply:
We should not go into the recursive loop again with a wrong name `fieldName`, this name is used as a look up key( case insensitive) for `ITEM` operator.

Reply:
Here the ```fieldName``` has meaning of result name of projection expression, when it's necessary to lookup field within struct (ITEM operator), then used ```literalString``` (value of ITEM's second operand).

Reply:
Maybe we should use a name pattern like `fieldName$0`, if it was in a deep recursive, the name may like `fieldName$0$0...`

Reply:
ok, I'll add the concatenation.

Reply:
@danny0405 done.